contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -168,47 +172,128 @@ type KMetadata interface {
 }
 ```
 
+## Contextual logging in Kubernetes
+
+Contextual logging builds on top of structured logging because the parameters
+for individual log calls are the same. The difference is that different
+functions need to be used:
+
+- `klog.ErrorS` -> `logger.Error`
+- `klog.InfoS` -> `logger.Info`
+- `klog.V().InfoS` -> `logger.V().Info`
+
+In all of these cases, `logger` is a `logr.Logger` instance. `klog.Logger` is
+an alias for that type. Determining where that instance comes from is the main
+challenge when migrating code to contextual logging.
+
+Several new klog functions help with that:
+
+- [`klog.FromContext`](https://pkg.go.dev/k8s.io/klog/v2#FromContext)
+- [`klog.Background`](https://pkg.go.dev/k8s.io/klog/v2#Background)
+- [`klog.TODO`](https://pkg.go.dev/k8s.io/klog/v2#TODO)
+
+The preferred approach is to retrieve the instance with `klog.FromContext` from
+a `ctx context` parameter. If a function or method does not have one, consider
+adding it. This API change then implies that all callers also need to be
+updated. If there are any `context.TODO` calls in the modified functions,
+replace with the new `ctx` parameter.
+
+In performance critical code it may be faster to add a `logger klog.Logger`
+parameter. This needs to be decided on a case-by-case basis.
+
+When such API changes trickle up to a unit test, then enable contextual logging
+with per-test output with
+[`ktesting`](https://pkg.go.dev/k8s.io/klog/v2@v2.60.1/ktesting):
+
+```go
+import (
+    "testing"
+
+    "k8s.io/klog/v2/ktesting"
+    _ "k8s.io/klog/v2/ktesting/init" # Add command line flags for ktesting.
+)
+
+func TestFoo(t *testing.T) {
+   _, ctx := ktesting.NewTestContext(t)
+   doSomething(ctx)
+}
+```
+
+If a logger instance is needed instead, then use:
+
+```go
+   logger, _ := ktesting.NewTestContext(t)
+```
+
+The KEP has further instructions about the [transition to contextual
+logging](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging#transition). It
+also lists several
+[pitfalls](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging#pitfalls-during-usage)
+that developers and reviewers need to be aware of.
+
 ## Migration
 
-1. Change log functions to structured equivalent
+1. Find code locations that need to be changed:
```

Comment:
I think that this tool would be better to verify the results than find code locations.

Reply:
> I think that this tool would be better to verify the results than find code locations.

Then how do you suggest that developers should find code locations instead? Using "git grep" easily leads to false positives. The logcheck tool is more precise.

> Could make it part of Verify log output?

That section is about something else (the output that gets generated, not how it gets generated). I can add a "Preventing regressions" section and there describe how we use `hack/logcheck.conf` to ensure that undesirable calls do not get added back into already converted packages.